gax/src/main/java/com/google/api/gax/tracing/TracedClientStreamingCallable.java

```diff
@@ -0,0 +1,168 @@
+ */
+@BetaApi("The surface for tracing is not stable and might change in the future")
+@InternalApi
+public class TracedClientStreamingCallable<RequestT, ResponseT>
+    extends ClientStreamingCallable<RequestT, ResponseT> {
+  private final ClientStreamingCallable<RequestT, ResponseT> innerCallable;
+  private final ApiTracerFactory tracerFactory;
+  private final SpanName spanName;
+
+  public TracedClientStreamingCallable(
+      @Nonnull ClientStreamingCallable<RequestT, ResponseT> innerCallable,
+      @Nonnull ApiTracerFactory tracerFactory,
+      @Nonnull SpanName spanName) {
+    this.tracerFactory = Preconditions.checkNotNull(tracerFactory, "tracerFactory can't be null");
+    this.spanName = Preconditions.checkNotNull(spanName, "spanName can't be null");
+    this.innerCallable = Preconditions.checkNotNull(innerCallable, "innerCallable can't be null");
+  }
+
+  @Override
+  public ApiStreamObserver<RequestT> clientStreamingCall(
+      ApiStreamObserver<ResponseT> responseObserver, ApiCallContext context) {
+
+    ApiTracer tracer = tracerFactory.newTracer(spanName);
+    context = context.withTracer(tracer);
+
+    // Shared state that allows the response observer to know that the error it received was
+    // triggered by a caller initiated cancellation.
+    AtomicReference<Throwable> cancellationCauseHolder = new AtomicReference<>(null);
+
+    try {
+      ApiStreamObserver<ResponseT> innerResponseObserver =
+          new TracedResponseObserver<>(tracer, responseObserver, cancellationCauseHolder);
+
+      ApiStreamObserver<RequestT> innerRequestObserver =
+          innerCallable.clientStreamingCall(innerResponseObserver, context);
+
+      return new TracedRequestObserver<>(tracer, innerRequestObserver, cancellationCauseHolder);
+    } catch (RuntimeException e) {
+      tracer.operationFailed(e);
+      throw e;
+    }
+  }
+
+  /** {@link ApiStreamObserver} wrapper to annotate the current trace with sent message. */
+  private static class TracedRequestObserver<RequestT> implements ApiStreamObserver<RequestT> {
+    private final ApiTracer tracer;
+    private final ApiStreamObserver<RequestT> innerObserver;
+    private final AtomicReference<Throwable> cancellationCauseHolder;
+
+    TracedRequestObserver(
+        @Nonnull ApiTracer tracer,
+        @Nonnull ApiStreamObserver<RequestT> innerObserver,
+        @Nonnull AtomicReference<Throwable> cancellationCauseHolder) {
+      this.tracer = Preconditions.checkNotNull(tracer, "tracer can't be null");
+      this.innerObserver = Preconditions.checkNotNull(innerObserver, "innerObserver can't be null");
+      this.cancellationCauseHolder =
+          Preconditions.checkNotNull(
+              cancellationCauseHolder, "cancellationCauseHolder can't be null");
+    }
+
+    @Override
+    public void onNext(RequestT value) {
+      tracer.requestSent();
+      innerObserver.onNext(value);
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      if (t == null) {
```

Comment:
So `onError` can be called only in case of cancelation? Why? Btw, please rename `t` to `throwable` or `error` or `exception`.

Why is it different from similar logic in `TraceResponseObserver` on line 153?

Reply:
This is the request stream half. Sending an error on the request stream will end up closing the response stream. So this is more similar to unary future cancellation then a TraceResponseObserver